### DISCUSSION TITLE
Resume after break

### DISCUSSION THREAD

**Discussion by bentoayr** [coin-or/rbfopt#39] [General]
Hello,

I am optimizing a *very* complicated function whose evaluation requires calling multiple programs via a connection to the cloud, etc.

Sometimes, connection breaks, which forces me to start all over.

Does rbfopt allow me to save the state of solving an optimization problem and resume after a fail?

Do you know of any black box optimizer that allows me to do that?

Thanks !

**Comment by gnannicini** [comment]
Yes, you can save the state.
You have to create your own script to create a RbfoptBlackBox and an RbfoptAlgorithm. After that, you can execute one iteration at a time "manually" by repeatedly running

`alg.optimize(pause_after_iters=1)`

where alg is your RbfoptAlgorithm object. The state can be saved with

`alg.save_to_file(filename)`

and there is a classmethod `load_from_file(filename)` if you want to restore.

**Reply by bentoayr** [reply]
Also, can the RbfoptAlgorithm be one of the existing algorithms provided by the package?

**Reply by gnannicini** [reply]
I don't understand this question. RbfoptAlgorithm is *the* algorithm provided by the package.

**Comment by bentoayr** [comment]
OK. This is very useful !

I'm new to python, is there some simple example with these ideas already
implemented ?

Thanks ,

J

**Comment by gnannicini** [comment]
I'm not aware of an example specifically for this, sorry. 
There is an example of loading the state from file here: https://github.com/coin-or/rbfopt/blob/master/bin/rbfopt_cl_interface.py

**Reply by bentoayr** [reply]
Hello again,

I tried to follow your instructions. I can save the state and load the state.
However, if I stop the optimization midway, and then I load the saved file using                 

`alg =  rbfopt.RbfoptAlgorithm.load_from_file("./saved_rbf_opt_state")`

to try to resume the optimization, when I proceed to iterate using 

`val, x, itercount, evalcount, fast_evalcount = alg.optimize(pause_after_iters = 1)`

the rbf just finishes without making any more iterations. 

To load the previous state, should I also call again the rbfopt.RbfoptUserBlackBox and call again the rbfopt.RbfoptSettings ?

**Comment by gnannicini** [comment]
When you resume, the algorithm object `alg` already stores the settings and black-box function.

I am not sure why you are seeing that behavior. It is supposed to work. I encourage you to try to debug a little bit to see if the black box is called at all, and if you can't get it to work, then please open an issue, providing more information and possibly a way to replicate the problem so I can look into it.

**Comment by bentoayr** [comment]
Fixed !
It was my lack of python knowledge that caused then problem.

fyi: I've checked several black-box solvers outthere, yours seems to be the
best so far. Congrats and thanks for then good work !
